### PR TITLE
add idempotency docs

### PR DESCRIPTION
**Pull request by mahendraHegde**
for https://github.com/novuhq/novu/issues/4469

### CODE REVIEW

**Review by unicodeveloper — APPROVED** [review]

**Comment by rifont** [comment]
@mahendraHegde Deliveroo have adopted the same Idempotency approach, albeit with slightly different naming. You may like to use them as a reference:
https://api-docs.deliveroo.com/docs/signature-idempotent-requests

**Comment by mahendraHegde** [comment]
@unicodeveloper  docs re readt to be reviewed
@rifont  i added the docs at path `additional-resources/idempotency` as it seems the std used in docs, so we might need to change `Link` response header in the api as well

File: additional-resources/idempotency.mdx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+When you make idempotent requests, the API stores the status code and response body from the initial request. If the same idempotency key is used in another request, the API replays the saved response, even for error responses like 400 and 500 errors.
+
+When a response is replayed, an `Idempotent-Replay: true` header is included in the response.
+
+## Validation for Idempotent Requests
+
+Idempotent requests are subject to the following validations:
+
+1. If the `<key>` exceeds 255 characters, the API will respond with a `Bad Request` error.
+2. If the request parameters do not match the initial request, the API returns an `Unprocessable Entity` error.
```
Comment (rifont) [inline additional-resources/idempotency.mdx:19]: We are only checking on the request body currently.

```suggestion
2. If the request body does not match the initial request, the API returns a `422 Unprocessable Entity` error.
```

File: additional-resources/idempotency.mdx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+When you make idempotent requests, the API stores the status code and response body from the initial request. If the same idempotency key is used in another request, the API replays the saved response, even for error responses like 400 and 500 errors.
+
+When a response is replayed, an `Idempotent-Replay: true` header is included in the response.
+
+## Validation for Idempotent Requests
+
+Idempotent requests are subject to the following validations:
+
+1. If the `<key>` exceeds 255 characters, the API will respond with a `Bad Request` error.
```
Comment (rifont) [inline additional-resources/idempotency.mdx:18]: Including the status code for quicker reference.

```suggestion
1. If the `<key>` exceeds 255 characters, the API will respond with a `400 Bad Request` error.
```

File: additional-resources/idempotency.mdx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+## Performing an Idempotent Request
+
+To make an idempotent request, include the `Idempotency-Key: <key>` header in your request. The idempotency key should be a unique client-generated value with sufficient entropy to prevent collisions. We recommend using V4 UUIDs as your idempotency keys.
+
+## Response Replay
+
+When you make idempotent requests, the API stores the status code and response body from the initial request. If the same idempotency key is used in another request, the API replays the saved response, even for error responses like 400 and 500 errors.
+
+When a response is replayed, an `Idempotent-Replay: true` header is included in the response.
```
Comment (rifont) [inline additional-resources/idempotency.mdx:12]: This is one of the implementation details we have different to Deliveroo.

```suggestion
When a response is replayed, an `Idempotency-Replay: true` header is included in the response.
```

File: additional-resources/idempotency.mdx [hunk 4]
```diff
@@ -0,0 +1,24 @@
+
+Novu supports retrying and concurrent requests for `POST` and `PATCH` APIs, ensuring the safety and reliability of your requests, particularly in scenarios where network issues or communication problems can introduce uncertainty into request outcomes.
+
+## Performing an Idempotent Request
+
+To make an idempotent request, include the `Idempotency-Key: <key>` header in your request. The idempotency key should be a unique client-generated value with sufficient entropy to prevent collisions. We recommend using V4 UUIDs as your idempotency keys.
+
+## Response Replay
+
+When you make idempotent requests, the API stores the status code and response body from the initial request. If the same idempotency key is used in another request, the API replays the saved response, even for error responses like 400 and 500 errors.
```
Comment (rifont) [inline additional-resources/idempotency.mdx:10]: Small change for readability and formality.

```suggestion
When you make idempotent requests, the API stores the status code and response body from the initial request. If the same idempotency key is used in another request, the API replays the saved response, including for client and server error responses like 400 and 500 errors.
```

File: additional-resources/idempotency.mdx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+
```
Comment (rifont) [inline additional-resources/idempotency.mdx:1]:
```suggestion
---
title: 'Idempotent Requests'
description: 'Learn how to retry API requests without performing the same operation twice'
---

```

File: additional-resources/idempotency.mdx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+When a response is replayed, an `Idempotent-Replay: true` header is included in the response.
+
+## Validation for Idempotent Requests
+
+Idempotent requests are subject to the following validations:
+
+1. If the `<key>` exceeds 255 characters, the API will respond with a `Bad Request` error.
+2. If the request parameters do not match the initial request, the API returns an `Unprocessable Entity` error.
+3. If the initial request is still being processed concurrently and hasn't yet responded, the API returns a `Conflict` error along with a `Retry-After` header.
```
Comment (rifont) [inline additional-resources/idempotency.mdx:20]:
```suggestion
3. If the initial request is still being processed concurrently and hasn't yet responded, the API returns a `409 Conflict` error along with a `Retry-After` header.
```

File: mint.json [hunk 7]
```diff
@@ -386,7 +386,8 @@
       "group": "Additional Resources",
       "pages": [
         "additional-resources/rate-limits",
-        "additional-resources/glossary"
+        "additional-resources/glossary",
+        "additional-resources/idempotency"
```
Comment (rifont) [inline mint.json:390]: Closer grouping of Rate Limits and Idempotency as they are related with retry behaviour.

```suggestion
        "additional-resources/idempotency",
        "additional-resources/glossary"
```

**Review by rifont — CHANGES_REQUESTED** [review]

File: additional-resources/idempotency.mdx [hunk 8]
```diff
@@ -0,0 +1,28 @@
+---
+title: 'Idempotent Requests'
+description: 'Learn how to retry API requests without performing the same operation twice'
+---
+
+Novu supports retrying and concurrent requests for `POST` and `PATCH` APIs, ensuring the safety and reliability of your requests, particularly in scenarios where network issues or communication problems can introduce uncertainty into request outcomes.
```
Comment (rifont) [inline additional-resources/idempotency.mdx:6]:
```suggestion
Novu supports retrying and concurrent requests for `POST` and `PATCH` operations, ensuring the safety and reliability of your requests, particularly in scenarios where network issues or communication problems can introduce uncertainty into request outcomes.
```

**Review by rifont — CHANGES_REQUESTED** [review]

File: additional-resources/idempotency.mdx [hunk 9]
```diff
@@ -0,0 +1,28 @@
+---
+title: 'Idempotent Requests'
+description: 'Learn how to retry API requests without performing the same operation twice'
+---
+
+Novu supports retrying and concurrent requests for `POST` and `PATCH` operations, ensuring the safety and reliability of your requests, particularly in scenarios where network issues or communication problems can introduce uncertainty into request outcomes.
```
Comment (rifont) [inline additional-resources/idempotency.mdx:6]: Adding context relevant to Novu.

```suggestion
Novu supports retrying and concurrent requests for `POST` and `PATCH` operations, ensuring the safety and reliability of your requests, particularly in scenarios where network issues or communication problems can introduce uncertainty into request outcomes. Idempotent requests are essential for business-critical applications, such as for ensuring only once delivery of your notifications.
```

**Review by rifont — CHANGES_REQUESTED** [review]

**Review by rifont — APPROVED** [review]

**Comment by rifont** [comment]
Great work @mahendraHegde ! Thanks for your contribution 🎉

**Pull request merged**